2015-Fall/lrusnac/CalculatorV2/CalculatorV2/CalculatorModel.swift

```diff
@@ -0,0 +1,107 @@
+//
+//  CalculatorModel.swift
+//  CalculatorV2
+//
+//  Created by Leonid Rusnac on 08/09/15.
+//  Copyright (c) 2015 Leonid Rusnac. All rights reserved.
+//
+
+import Foundation
+
+class CalculatorModel {
+
+    private enum Op: Printable{
+        case Operand(Double)
+        case UnaryOperator(String, Double -> Double)
+        case BinaryOperator(String, (Double, Double) -> Double)
+        case Constant(String, Double)
+        
+        // equivalent of toString in java
+        var description: String {
+            get {
+                switch self {
+                case .Operand(let operand):
+                    return "\(operand)"
+                case .UnaryOperator(let operatorSymbol, _):
+                    return operatorSymbol
+                case .BinaryOperator(let operatorSymbol, _):
+                    return operatorSymbol
+                case .Constant(let operatorSymbol, _):
+                    return operatorSymbol
+                }
+            }
+        }
+    }
+    
+    private var ops = [Op]()
+    private var knownOps = [String:Op]()
+    
+    
+    init() {
+        func learnOp(op: Op) {
+            knownOps[op.description] = op
+        }
+        
+        learnOp(Op.BinaryOperator("+", +))
+        learnOp(Op.BinaryOperator("×", *))
+        learnOp(Op.BinaryOperator("÷", {$1/$0}))
+        learnOp(Op.BinaryOperator("-", {$1-$0}))
+        
+        learnOp(Op.UnaryOperator("√", sqrt))
+        learnOp(Op.UnaryOperator("sin", sin))
+        learnOp(Op.UnaryOperator("cos", cos))
+        
+        learnOp(Op.Constant("π", M_PI))
+    }
+    
+    func pushNumber(number: Double) {
+        ops.append(Op.Operand(number))
+    }
+    
+    func pushOpNotNumber(keyString: String) {
+        if let con = knownOps[keyString] {
+            ops.append(con)
+        }
+    }
+    
+    func clear() {
+        ops.removeAll(keepCapacity: false)
```

Comment:
I raised with pointers in C :D 
So I always have in mind to clean after me, even if there is a reference count and cleans automatically.

I put the value just to remember it is an option, I know that is useless but you don't really have this option in every programming language.